File: Desktop/GB/backend/populate_mock_data.py
#!/usr/bin/env python3
"""
Script to populate the database with mock data for testing and demonstration
"""

import os
import sys
from datetime import datetime, date, timedelta
from random import randint, choice, uniform

# Add the backend directory to the path
sys.path.append(os.path.dirname(os.path.abspath(__file__)))

from app import app
from database import db
from models.cattle import Cattle
from models.milk_production import MilkProduction
from models.feeding import Feeding
from models.expenses import Expenses
from models.revenue import Revenue

# Mock data constants
BREEDS = ['Holstein', 'Jersey', 'Angus', 'Hereford', 'Brahman', 'Simmental', 'Charolais']
HEALTH_STATUSES = ['Healthy', 'Sick', 'Injured', 'Pregnant', 'Recovering']
CURRENT_STATUSES = ['Active', 'Sold', 'Deceased', 'Quarantined']
LOCATIONS = ['Barn A', 'Barn B', 'Pasture 1', 'Pasture 2', 'Quarantine Area']
FEED_TYPES = ['Hay', 'Corn Silage', 'Barley', 'Wheat', 'Alfalfa', 'Grass Pellets', 'Protein Supplement']
EXPENSE_CATEGORIES = ['Feed', 'Veterinary', 'Equipment', 'Maintenance', 'Utilities', 'Labor', 'Insurance']
REVENUE_SOURCES = ['Milk Sales', 'Cattle Sales', 'Breeding Services', 'Manure Sales']

def create_mock_cattle():
    """Create mock cattle records"""
    cattle_data = []
    
    for i in range(1, 16):  # Create 15 cattle
        # Generate random dates
        birth_date = date.today() - timedelta(days=randint(365, 2190))  # 1-6 years old
        purchase_date = birth_date + timedelta(days=randint(30, 365))  # Purchased sometime after birth
        
        cattle = Cattle(
            tag_number=f"GB{i:04d}",
            name=choice([
                'Bella', 'Daisy', 'Molly', 'Bessie', 'Luna', 'Rosie', 'Stella', 'Ruby',
                'Max', 'Charlie', 'Duke', 'Thunder', 'Rocky', 'Bruno', 'Atlas'
            ]) + f" {i}",
            breed=choice(BREEDS),
            date_of_birth=birth_date,
            gender=choice(['Male', 'Female']),
            weight=uniform(400, 800),
            health_status=choice(HEALTH_STATUSES) if randint(1, 10) <= 2 else 'Healthy',  # 20% chance of non-healthy
            location=choice(LOCATIONS),
            purchase_date=purchase_date,
            purchase_price=uniform(800, 2500),
            current_status=choice(CURRENT_STATUSES) if randint(1, 10) <= 1 else 'Active',  # 10% chance of non-active
            notes=choice([
                'Excellent milk producer',
                'Good temperament',
                'Requires special attention',
                'Regular health checks needed',
                'High breeding value',
                None
            ])
        )
        cattle_data.append(cattle)
    
    return cattle_data

def create_mock_milk_production(cattle_list):
    """Create mock milk production records"""
    milk_records = []
    
    for cattle in cattle_list:
        if cattle.gender == 'Female' and cattle.current_status == 'Active':
            # Generate milk records for the last 30 days
            for i in range(30):
                record_date = date.today() - timedelta(days=i)
                quantity = uniform(15, 35)  # 15-35 liters per day
                
                milk_record = MilkProduction(
                    cattle_id=cattle.id,
                    date_recorded=record_date,
                    quantity_liters=quantity,
                    quality_score=uniform(7.0, 10.0),
                    notes=choice([
                        'Morning milking',
                        'Evening milking',
                        'Good quality',
                        'Lower than usual',
                        'Excellent production',
                        None
                    ])
                )
                milk_records.append(milk_record)
    
    return milk_records

def create_mock_feeding(cattle_list):
    """Create mock feeding records"""
    feeding_records = []
    
    for cattle in cattle_list:
        if cattle.current_status == 'Active':
            # Generate feeding records for the last 7 days
            for i in range(7):
                record_date = date.today() - timedelta(days=i)
                feed_type = choice(FEED_TYPES)
                quantity = uniform(5, 15)  # 5-15 kg per day
                cost_per_unit = uniform(0.5, 2.0)  # Cost per kg
                
                feeding_record = Feeding(
                    cattle_id=cattle.id,
                    date_recorded=record_date,
                    feed_type=feed_type,
                    quantity_kg=quantity,
                    cost_per_unit=cost_per_unit,
                    total_cost=quantity * cost_per_unit,
                    supplier=choice(['FeedCorp', 'AgriSupply', 'FarmFresh', 'LocalFarm']),
                    notes=choice([
                        'Regular feeding',
                        'Extra nutrition',
                        'Special diet',
                        'Reduced portion',
                        None
                    ])
                )
                feeding_records.append(feeding_record)
    
    return feeding_records

def create_mock_expenses():
    """Create mock expense records"""
    expenses = []
    
    for i in range(20):  # Create 20 expense records
        expense_date = date.today() - timedelta(days=randint(1, 90))
        category = choice(EXPENSE_CATEGORIES)
        
        # Generate realistic amounts based on category
        if category == 'Feed':
            amount = uniform(200, 800)
            description = f"{choice(FEED_TYPES)} purchase"
        elif category == 'Veterinary':
            amount = uniform(50, 300)
            description = choice(['Vaccination', 'Health check', 'Treatment', 'Emergency visit'])
        elif category == 'Equipment':
            amount = uniform(100, 1500)
            description = choice(['Milking equipment', 'Feeding tools', 'Maintenance tools'])
        elif category == 'Maintenance':
            amount = uniform(50, 400)
            description = choice(['Barn repair', 'Fence maintenance', 'Equipment repair'])
        elif category == 'Utilities':
            amount = uniform(100, 300)
            description = choice(['Electricity', 'Water', 'Gas'])
        elif category == 'Labor':
            amount = uniform(300, 800)
            description = choice(['Farm worker salary', 'Veterinarian fee', 'Consultant fee'])
        else:  # Insurance
            amount = uniform(200, 600)
            description = 'Insurance premium'
        
        expense = Expenses(
            date_recorded=expense_date,
            category=category,
            description=description,
            amount=amount,
            supplier=choice(['FeedCorp', 'AgriSupply', 'VetClinic', 'LocalSupplier', 'UtilityCorp']),
            receipt_number=f"REC{randint(1000, 9999)}",
            notes=choice([
                'Paid in full',
                'Partial payment',
                'Invoice pending',
                'Urgent expense',
                None
            ])
        )
        expenses.append(expense)
    
    return expenses

def create_mock_revenue():
    """Create mock revenue records"""
    revenues = []
    
    for i in range(15):  # Create 15 revenue records
        revenue_date = date.today() - timedelta(days=randint(1, 60))
        source = choice(REVENUE_SOURCES)
        
        # Generate realistic amounts based on source
        if source == 'Milk Sales':
            amount = uniform(500, 1200)
            description = f"Milk sales - {randint(100, 300)} liters"
        elif source == 'Cattle Sales':
            amount = uniform(1000, 3000)
            description = f"Cattle sale - {choice(BREEDS)} {choice(['Bull', 'Cow', 'Heifer'])}"
        elif source == 'Breeding Services':
            amount = uniform(100, 400)
            description = "Breeding service fee"
        else:  # Manure Sales
            amount = uniform(50, 200)
            description = "Organic manure sale"
        
        revenue = Revenue(
            date_recorded=revenue_date,
            source=source,
            description=description,
            amount=amount,
            notes=choice([
                'Payment received',
                'Pending payment',
                'Partial payment',
                'Full payment',
                None
            ])
        )
        revenues.append(revenue)
    
    return revenues

def populate_database():
    """Main function to populate the database with mock data"""
    with app.app_context():
        try:
            # Clear existing data
            print("Clearing existing data...")
            db.drop_all()
            db.create_all()
            
            # Create mock cattle
            print("Creating mock cattle...")
            cattle_list = create_mock_cattle()
            db.session.add_all(cattle_list)
            db.session.commit()
            
            # Refresh cattle objects to get their IDs
            cattle_list = Cattle.query.all()
            
            # Create mock milk production records
            print("Creating mock milk production records...")
            milk_records = create_mock_milk_production(cattle_list)
            db.session.add_all(milk_records)
            db.session.commit()
            
            # Create mock feeding records
            print("Creating mock feeding records...")
            feeding_records = create_mock_feeding(cattle_list)
            db.session.add_all(feeding_records)
            db.session.commit()
            
            # Create mock expenses
            print("Creating mock expenses...")
            expenses = create_mock_expenses()
            db.session.add_all(expenses)
            db.session.commit()
            
            # Create mock revenues
            print("Creating mock revenues...")
            revenues = create_mock_revenue()
            db.session.add_all(revenues)
            db.session.commit()
            
            print("\n✅ Mock data populated successfully!")
            print(f"Created:")
            print(f"  - {len(cattle_list)} cattle records")
            print(f"  - {len(milk_records)} milk production records")
            print(f"  - {len(feeding_records)} feeding records")
            print(f"  - {len(expenses)} expense records")
            print(f"  - {len(revenues)} revenue records")
            
        except Exception as e:
            print(f"❌ Error populating database: {e}")
            db.session.rollback()
            raise

if __name__ == '__main__':
    populate_database()

File: Desktop/GB/cattle-management-mobile/backend-mongo/scripts/populate-mock-data.js
const mongoose = require('mongoose');
require('dotenv').config();

// Import models
const Cattle = require('../models/Cattle');
const MilkProduction = require('../models/MilkProduction');
const Feeding = require('../models/Feeding');
const Expense = require('../models/Expense');
const Revenue = require('../models/Revenue');

// Mock data constants
const BREEDS = ['Holstein', 'Jersey', 'Angus', 'Hereford', 'Brahman', 'Simmental', 'Charolais'];
const HEALTH_STATUSES = ['Healthy', 'Sick', 'Injured', 'Pregnant', 'Recovering'];
const CURRENT_STATUSES = ['Active', 'Sold', 'Deceased', 'Quarantined'];
const LOCATIONS = ['Barn A', 'Barn B', 'Pasture 1', 'Pasture 2', 'Quarantine Area'];
const FEED_TYPES = ['Hay', 'Corn Silage', 'Barley', 'Wheat', 'Alfalfa', 'Grass Pellets', 'Protein Supplement'];
const EXPENSE_CATEGORIES = ['Feed', 'Veterinary', 'Equipment', 'Maintenance', 'Utilities', 'Labor', 'Insurance'];
const REVENUE_SOURCES = ['Milk Sales', 'Cattle Sales', 'Breeding Services', 'Manure Sales'];

// Helper functions
const randomChoice = (array) => array[Math.floor(Math.random() * array.length)];
const randomFloat = (min, max) => Math.random() * (max - min) + min;
const randomInt = (min, max) => Math.floor(Math.random() * (max - min + 1)) + min;
const randomDate = (start, end) => new Date(start.getTime() + Math.random() * (end.getTime() - start.getTime()));

async function createMockCattle() {
  console.log('Creating mock cattle...');
  const cattle = [];
  
  for (let i = 1; i <= 15; i++) {
    const dateOfBirth = randomDate(new Date(2020, 0, 1), new Date(2023, 11, 31));
    const purchaseDate = randomDate(new Date(2023, 0, 1), new Date(2024, 11, 31));
    
    const cattleData = {
      tag_number: `GB${String(i).padStart(4, '0')}`,
      name: `${randomChoice(['Daisy', 'Bella', 'Luna', 'Rosie', 'Molly', 'Ruby', 'Stella', 'Coco', 'Penny', 'Ginger'])} ${i}`,
      breed: randomChoice(BREEDS),
      date_of_birth: dateOfBirth,
      gender: randomChoice(['Male', 'Female']),
      weight: randomFloat(400, 800),
      health_status: randomChoice(HEALTH_STATUSES),
      location: randomChoice(LOCATIONS),
      purchase_date: purchaseDate,
      purchase_price: randomFloat(1000, 2500),
      current_status: randomChoice(CURRENT_STATUSES.filter(s => s !== 'Deceased')), // Mostly active cattle
      notes: randomChoice([
        'High milk producer',
        'Good breeding stock',
        'Excellent health record',
        'Strong genetics',
        'High breeding value',
        null
      ])
    };
    
    cattle.push(cattleData);
  }
  
  return await Cattle.insertMany(cattle);
}

async function createMockMilkProduction(cattleList) {
  console.log('Creating mock milk production records...');
  const milkRecords = [];
  
  // Create records for the last 30 days
  const endDate = new Date();
  const startDate = new Date();
  startDate.setDate(startDate.getDate() - 30);
  
  for (const cattle of cattleList) {
    // Skip if cattle is not active or is male
    if (cattle.current_status !== 'Active' || cattle.gender === 'Male') continue;
    
    // Create 1-2 records per day for each cattle
    for (let d = new Date(startDate); d <= endDate; d.setDate(d.getDate() + 1)) {
      const recordsPerDay = randomInt(1, 2);
      
      for (let r = 0; r < recordsPerDay; r++) {
        const recordDate = new Date(d);
        recordDate.setHours(randomInt(5, 18), randomInt(0, 59), 0, 0);
        
        milkRecords.push({
          cattle_id: cattle._id,
          date_recorded: recordDate,
          quantity_liters: randomFloat(15, 35), // Liters per session
          quality_score: randomFloat(7, 10),
          notes: randomChoice([
            'Good quality milk',
            'Excellent production',
            'Normal milking session',
            'High fat content',
            null
          ])
        });
      }
    }
  }
  
  return await MilkProduction.insertMany(milkRecords);
}

async function createMockFeeding(cattleList) {
  console.log('Creating mock feeding records...');
  const feedingRecords = [];
  
  // Create records for the last 30 days
  const endDate = new Date();
  const startDate = new Date();
  startDate.setDate(startDate.getDate() - 30);
  
  for (const cattle of cattleList) {
    // Skip if cattle is not active
    if (cattle.current_status !== 'Active') continue;
    
    // Create 2-3 feeding records per day for each cattle
    for (let d = new Date(startDate); d <= endDate; d.setDate(d.getDate() + 1)) {
      const feedingsPerDay = randomInt(2, 3);
      
      for (let f = 0; f < feedingsPerDay; f++) {
        const feedDate = new Date(d);
        feedDate.setHours(randomInt(6, 20), randomInt(0, 59), 0, 0);
        
        const feedType = randomChoice(FEED_TYPES);
        const quantity = randomFloat(5, 15); // kg
        const costPerUnit = randomFloat(0.5, 2.5); // per kg
        
        feedingRecords.push({
          cattle_id: cattle._id,
          date_recorded: feedDate,
          feed_type: feedType,
          quantity_kg: quantity,
          cost_per_unit: costPerUnit,
          total_cost: quantity * costPerUnit,
          supplier: randomChoice(['Farm Supply Co', 'Green Valley Feeds', 'Country Feed Store', 'Local Farmer', null]),
          notes: randomChoice([
            'Regular feeding',
            'Extra nutrition needed',
            'High quality feed',
            'Bulk purchase discount',
            null
          ])
        });
      }
    }
  }
  
  return await Feeding.insertMany(feedingRecords);
}

async function createMockExpenses() {
  console.log('Creating mock expenses...');
  const expenses = [];
  
  // Create expenses for the last 90 days
  const endDate = new Date();
  const startDate = new Date();
  startDate.setDate(startDate.getDate() - 90);
  
  for (let i = 0; i < 25; i++) {
    const expenseDate = randomDate(startDate, endDate);
    const category = randomChoice(EXPENSE_CATEGORIES);
    
    let amount, description;
    switch (category) {
      case 'Feed':
        amount = randomFloat(200, 800);
        description = `${randomChoice(FEED_TYPES)} purchase`;
        break;
      case 'Veterinary':
        amount = randomFloat(50, 300);
        description = randomChoice(['Vaccination', 'Health checkup', 'Treatment', 'Emergency care']);
        break;
      case 'Equipment':
        amount = randomFloat(100, 1500);
        description = randomChoice(['Milking equipment', 'Feeding tools', 'Maintenance tools', 'Farm machinery']);
        break;
      case 'Maintenance':
        amount = randomFloat(75, 500);
        description = randomChoice(['Barn repair', 'Fence maintenance', 'Equipment service', 'Facility upgrade']);
        break;
      case 'Utilities':
        amount = randomFloat(150, 400);
        description = randomChoice(['Electricity bill', 'Water bill', 'Gas bill', 'Internet/Phone']);
        break;
      case 'Labor':
        amount = randomFloat(300, 1200);
        description = randomChoice(['Farm worker salary', 'Veterinarian fee', 'Consultant fee', 'Temporary help']);
        break;
      case 'Insurance':
        amount = randomFloat(200, 600);
        description = 'Insurance premium';
        break;
      default:
        amount = randomFloat(50, 500);
        description = 'General expense';
    }
    
    expenses.push({
      date_recorded: expenseDate,
      category: category,
      description: description,
      amount: amount,
      supplier: randomChoice(['Farm Supply Co', 'Local Vendor', 'Service Provider', 'Equipment Dealer', null]),
      receipt_number: `RCP${randomInt(1000, 9999)}`,
      notes: randomChoice([
        'Regular expense',
        'Urgent purchase',
        'Bulk discount applied',
        'Emergency expense',
        'Planned purchase',
        null
      ])
    });
  }
  
  return await Expense.insertMany(expenses);
}

async function createMockRevenue() {
  console.log('Creating mock revenue records...');
  const revenues = [];
  
  // Create revenue for the last 60 days
  const endDate = new Date();
  const startDate = new Date();
  startDate.setDate(startDate.getDate() - 60);
  
  for (let i = 0; i < 20; i++) {
    const revenueDate = randomDate(startDate, endDate);
    const source = randomChoice(REVENUE_SOURCES);
    
    let amount, description;
    switch (source) {
      case 'Milk Sales':
        amount = randomFloat(500, 2000);
        description = `Milk sales - ${randomInt(100, 500)} liters`;
        break;
      case 'Cattle Sales':
        amount = randomFloat(1500, 3500);
        description = `Cattle sale - ${randomChoice(['Bull', 'Cow', 'Heifer'])}`;
        break;
      case 'Breeding Services':
        amount = randomFloat(200, 800);
        description = 'Breeding service fee';
        break;
      case 'Manure Sales':
        amount = randomFloat(100, 400);
        description = 'Organic manure sales';
        break;
      default:
        amount = randomFloat(100, 1000);
        description = 'Other farm income';
    }
    
    revenues.push({
      date_recorded: revenueDate,
      source: source,
      description: description,
      amount: amount,
      notes: randomChoice([
        'Regular income',
        'Premium price received',
        'Bulk sale',
        'Contract sale',
        'Spot market sale',
        null
      ])
    });
  }
  
  return await Revenue.insertMany(revenues);
}

async function populateDatabase() {
  try {
    console.log('Connecting to MongoDB...');
    await mongoose.connect(process.env.MONGODB_URI);
    console.log('Connected to MongoDB');
    
    // Clear existing data
    console.log('Clearing existing data...');
    await Promise.all([
      Cattle.deleteMany({}),
      MilkProduction.deleteMany({}),
      Feeding.deleteMany({}),
      Expense.deleteMany({}),
      Revenue.deleteMany({})
    ]);
    
    // Create mock data
    const cattle = await createMockCattle();
    const milkRecords = await createMockMilkProduction(cattle);
    const feedingRecords = await createMockFeeding(cattle);
    const expenses = await createMockExpenses();
    const revenues = await createMockRevenue();
    
    console.log('\n✅ Mock data populated successfully!');
    console.log('Created:');
    console.log(`  - ${cattle.length} cattle records`);
    console.log(`  - ${milkRecords.length} milk production records`);
    console.log(`  - ${feedingRecords.length} feeding records`);
    console.log(`  - ${expenses.length} expense records`);
    console.log(`  - ${revenues.length} revenue records`);
    
  } catch (error) {
    console.error('Error populating database:', error);
  } finally {
    await mongoose.connection.close();
    console.log('Database connection closed');
  }
}

// Run the population script
if (require.main === module) {
  populateDatabase();
}

module.exports = { populateDatabase };
